woocommerce/google-listings-and-ads
# UX Improvement on Google Merchant Center section in Setup MC

### Changes proposed in this Pull Request:

Part of #1065.

This PR contains UX improvements on the Google Merchant Center (GMC) section in Setup MC flow.

- Users will start with either "Create account" flow or "Connect account" flow.
- If users do not have an existing GMC account, they will start with "Create account" flow, and go through the following UI: Create account card > Terms modal > "Creating" card > Reclaim URL card (if there is a need to reclaim URL) > Connected card.
- If users do have an existing GMC account, they will start with "Connect account" flow, and go through the following UI: Connect account card > Switch URL card (if there is a need to switch URL) > Reclaim URL card (if there is a need to reclaim URL) > Connected card.
    - In "Connect account card", users can choose to create a new GMC account. If there is an existing GMC account with matching domain, we will display a warning modal. Clicking "Yes" in the warning modal will display the terms modal for creating a new GMC account.
- When there is a reclaim URL error, instead of showing a separate "reclaim URL fail" card, we display the error using an error notice within the reclaim URL card. This means that we no longer need the `ReclaimUrlFailCard` component.
- After GMC account is connected, we now allow the users to switch to another GMC account.

Other notable technical changes:

- Make `useApiFetchCallback` support browser offline scenario. Prior to this, if browser is offline and apiFetch is triggered, the `loading` state will continue to be `true` even when there is a "browser is offline" error. Now, the `loading` state will become false.
- Refactor `useExistingGoogleMCAccounts` to make it use `useAppSelectDispatch` internally, instead of its own custom code with `isResolving` which is not an accurate way to indicate "loading" state anyway.

Note: Currently the cards do not show GMC account name and domain. That would be done in a separate PR later, pending for back-end API changes. We are primarily changing the UI flow here.

### Screenshots:

**Create Account UI**, when the connected Google account does not have any GMC accounts:

![image](https://user-images.githubusercontent.com/417342/141791737-21999916-d4f5-443a-bc7d-1454f8a0d5fa.png)

Clicking on the "Create account" button would open the terms agreement modal, as per existing behavior.

**Connect Account UI**, when the connected Google account has at least one Google Merchant Center account:

![image](https://user-images.githubusercontent.com/417342/141792075-74c2a37d-9df3-42a7-a755-17a13dd12df9.png)

Clicking on the "Or, create a new Merchant Center account" text button would open this **warning modal** if users already have a GMC account with domain that matches their current site domain:

![image](https://user-images.githubusercontent.com/417342/141792968-0241c493-0942-4174-9f5c-40f2c24ea081.png)

If the users do not have a GMC account with matching domain, then they won't see the above warning modal, and they will go straight to the terms agreement modal.

Clicking on the "Yes, I want a new account" will open the terms agreement modal (note that this modal is the same existing modal prior to this PR, and we are just reusing it in this new flow): 

![image](https://user-images.githubusercontent.com/417342/141793324-3cc09fec-8691-43ee-bffb-475fb6be5491.png)

Ticking the checkbox and clicking the "Create account" button would then proceed with the "Creating..." card below (this is the same existing card, we are just reusing it in this new flow):

![image](https://user-images.githubusercontent.com/417342/141793527-ba988805-8502-48eb-b6da-3f70a850c1ac.png)

**Switch URL Card** (note that "switch URL" case would only happen when you are connecting an account, and not when you are creating a new account):

![image](https://user-images.githubusercontent.com/417342/141795961-ef9a37be-6391-4ac2-996d-2e7e17de3c8a.png)

Clicking on the "Switch Account" button will bring you back to the "Connect Account UI".

**Reclaim URL Card**: 

![image](https://user-images.githubusercontent.com/417342/141796322-606e4c4b-f1b3-4012-bf32-dd02d43d06b2.png)

Clicking on the "Switch Account" button will bring you to the "Connect Account UI". This is true even when you are coming from "Create Account UI", because by this time your new GMC account would have been created, and when you go to "Connect Account UI", you should see the new GMC account in the account selection drop down list.

**Connected Account Card**: 

![image](https://user-images.githubusercontent.com/417342/142767953-9d8ca496-9185-4308-830d-38af3b258d44.png)

Clicking on the "Or, connect to a different Google Merchant Center account" will disconnect your account and bring you to the "Connect Account UI".

### Detailed test instructions:

Test note: 

- If you want to simulate the list of GMC accounts, you can manually modify `useExistingGoogleMCAccounts` so that it returns a list of GMC account objects. 
- If you want to trigger the warning modal below, you can manually modify `getMatchingDomainAccount` so that it returns a GMC account object.

1. Use a Google account that does not have any GMC account.
2. Go to Setup MC page. You should see the Create account card.
3. Go through the account creation process until you have an account connected. The UI should be as per screenshots above.
4. In the Connected account card, click on "connect to a different Google Merchant Center account". Your GMC account will be disconnected, the page will refetch your list of GMC accounts, and you should see the Connect account card.
5. Proceed to do account connection until you have an account connected. The UI should be as per screenshots above.
6. In the Connected account card, click on "connect to a different Google Merchant Center account". Your GMC account will be disconnected, the page will refetch your list of GMC accounts, and you should see the Connect account card.
7. In the Connect account card, click on "create a new Merchant Center account". You should see a warning modal.
8. Click "Yes". You should then see the terms modal.
9. Agree with the terms and proceed until you have a new account created and connected.



### Changelog entry

> Update - UX improvements on Google Merchant Center section in Setup MC flow.
